### PR TITLE
[hs-ledger-bindings tests] Read the main dalf from the DAR manifest

### PR DESCRIPTION
**Pull request by cocreature**
Previously, we got the package id by listing all packages and taking
the first. That is not a valid assumption and broke with a PR. This PR
changes the tests to use the DAR reader that we already used for
visualization (now factored into a separate library) and reads the
main dalf using that.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/daml-lf-reader/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+load("//bazel_tools:haskell.bzl", "da_haskell_library")
+
+da_haskell_library(
```
Comment (neil-da) [inline compiler/daml-lf-reader/BUILD.bazel]: Why not just shove this in the daml-lf-ast library? And perhaps rename that daml-lf

File: compiler/daml-lf-reader/src/DA/Daml/LF/Reader.hs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+import Codec.Archive.Zip
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.ByteString.Lazy.UTF8 as UTF8
+import qualified Data.HashMap.Strict as Map
+import Data.List.Extra
+import System.FilePath
+
+data Manifest = Manifest { mainDalf :: FilePath , dalfs :: [FilePath] } deriving (Show)
+data ManifestData = ManifestData { mainDalfContent :: BSL.ByteString , dalfsCotent :: [BSL.ByteString] } deriving (Show)
```
Comment (neil-da) [inline compiler/daml-lf-reader/src/DA/Daml/LF/Reader.hs]: I'd have split this into separate lines for the fields. Otherwise its hard to write comments.

Comment (neil-da) [inline compiler/daml-lf-reader/src/DA/Daml/LF/Reader.hs]: dalfsCotent => dalfsContent

Comment (neil-da) [inline compiler/daml-lf-reader/src/DA/Daml/LF/Reader.hs]: I wonder if we should have `Manifest a` where `a` is either `FilePath` or `BSL.ByteString`?

File: compiler/daml-lf-reader/src/DA/Daml/LF/Reader.hs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+lineToKeyValue :: String -> (String, String)
+lineToKeyValue line = case splitOn ":" line of
+    [l, r] -> (trim l , trim r)
+    _ -> ("malformed", "malformed")
```
Comment (neil-da) [inline compiler/daml-lf-reader/src/DA/Daml/LF/Reader.hs]: Not a fan of sneaking `malformed` into places - you're one day going to get an error that just talks about malformed... Perhaps just silently drop them with a Nothing?

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs [hunk 4]
```diff
@@ -40,12 +41,15 @@ sandboxProcess SandboxSpec{dar} = do
 startSandboxProcess :: SandboxSpec -> IO (ProcessHandle,Maybe Handle)
 startSandboxProcess spec = do
     processRecord <- sandboxProcess spec
-    (_,hOutOpt,_,proh) <-
+    (_,hOutOpt,Just hErrOpt,proh) <-
         createProcess processRecord {
         std_out = CreatePipe,
         std_err = CreatePipe, -- Question: ought the pipe to be drained?
```
Comment (neil-da) [inline language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: Why make this CreatePipe then drain it to stdout? Just pass a handle to std_err.

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs [hunk 5]
```diff
@@ -40,12 +41,15 @@ sandboxProcess SandboxSpec{dar} = do
...
         create_group = True  -- To avoid sending INT to ourself
         }
+    _ <- forkIO $ forever $ do
+        line <- hGetLine hErrOpt
+        print line
```
Comment (neil-da) [inline language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: This is odd. print and not putStrLn? Sending to stdout and not stderr?

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs [hunk 6]
```diff
@@ -77,6 +81,7 @@ getListeningLine :: Handle -> IO String
 getListeningLine h = loop where
     loop = do
         line <- hGetLine h
+        print line
```
Comment (neil-da) [inline language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: Debugging?

File: compiler/daml-lf-reader/BUILD.bazel (hunk 1 above)
Comment (neil-da) [reply compiler/daml-lf-reader/BUILD.bazel]: I think the proto generated vs hand written split makes sense, but from then on, i just want a daml-lf library. I don't feel too strongly though.

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs (hunk 5 above)
Comment (cocreature) [reply language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: oops that was only for debugging

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs (hunk 6 above)
Comment (cocreature) [reply language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: oops that what I get for trying to fix bugs quickly

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs (hunk 4 above)
Comment (cocreature) [reply language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: oops debugging :)

File: compiler/daml-lf-reader/BUILD.bazel (hunk 1 above)
Comment (cocreature) [reply compiler/daml-lf-reader/BUILD.bazel]: We already have `daml-lf-proto`, `daml-lf-ast`, and `daml-lf-tools`. I don’t really care where it ends up but given that we have a PR blocked on this fix, I would like to merge it without merging it for now.

File: compiler/daml-lf-reader/src/DA/Daml/LF/Reader.hs (hunk 2 above)
Comment (cocreature) [reply compiler/daml-lf-reader/src/DA/Daml/LF/Reader.hs]: changed

Comment (cocreature) [reply compiler/daml-lf-reader/src/DA/Daml/LF/Reader.hs]: Sounds reasonable to me but let’s change it in a separate PR.

File: compiler/daml-lf-reader/src/DA/Daml/LF/Reader.hs (hunk 3 above)
Comment (cocreature) [reply compiler/daml-lf-reader/src/DA/Daml/LF/Reader.hs]: I’ve changed it to a call to `error` for now, I’m not a fan of silently dropping on parse failures.

File: compiler/daml-lf-reader/BUILD.bazel (hunk 1 above)
Comment (neil-da) [reply compiler/daml-lf-reader/BUILD.bazel]: Certainly

File: compiler/daml-lf-reader/src/DA/Daml/LF/Reader.hs (hunk 2 above)
Comment (neil-da) [reply compiler/daml-lf-reader/src/DA/Daml/LF/Reader.hs]: 👍

**Comment by cocreature** [comment]
Removed the debugging output and some minor cleanup so this should be good to merge. I don’t feel strongly about any of this so happy to refactor and move things around afterwards.

**Comment by neil-da** [comment]
Agreed, happy to merge as is, especially given other things depend on it.

**Pull request merged**